compiler/lcalc/compile_without_exceptions.ml

```diff
@@ -34,513 +34,625 @@ module A = Ast
+      (fun vars ->
+        Ast.monad_return ~mark
+          (Expr.eapp
+             (Expr.eop (trans_op Op.Filter) tys opmark)
+             (ListLabels.map vars ~f:(fun v -> Expr.evar v m))
+             mark))
+      [f'; trans ctx l]
+      ~mark
+  | EApp { f = EOp { op = Op.Filter as op; _ }, _; _ }
+  | EApp { f = EOp { op = Op.Map as op; _ }, _; _ }
+  | EApp { f = EOp { op = Op.Fold as op; _ }, _; _ }
+  | EApp { f = EOp { op = Op.Reduce as op; _ }, _; _ } ->
+    (* Cannot happend: list operator must be fully determined *)
+    Errors.raise_internal_error
+      "List operator %a was not fully determined: some partial evaluation was \
+       found while compiling."
+      Print.operator op
+  | EApp { f = EOp { op; tys }, opmark; args } ->
+    let res =
+      Ast.monad_mmap
+        (Expr.eop (trans_op op) tys opmark)
+        (List.map (trans ctx) args)
+        ~mark
+    in
+    res
+  | EMatch { name; e; cases } ->
+    let cases =
+      EnumConstructor.MapLabels.map cases ~f:(fun case ->
+          match Marked.unmark case with
+          | EAbs { binder; tys } ->
+            let vars, body = Bindlib.unmbind binder in
+            let ctx' =
+              ArrayLabels.fold_right vars ~init:ctx ~f:(fun var ->
+                  Var.Map.add var
+                    {
+                      info_pure = true;
+                      is_scope = false;
+                      var = Var.translate var;
+                    })
+            in
+            let binder =
+              Expr.bind (Array.map Var.translate vars) (trans ctx' body)
+            in
+            Expr.eabs binder tys m
+          | _ -> assert false)
+    in
+    Ast.monad_bind_cont
+      (fun e -> Expr.ematch (Expr.evar e m) name cases m)
+      (trans ctx e) ~mark
+  | EArray args ->
+    Ast.monad_mbind_cont ~mark
+      (fun vars ->
+        Ast.monad_return ~mark
+          (Expr.earray
+             (List.map (fun v -> Ast.monad_return ~mark (Expr.evar v m)) vars)
+             mark))
+      (List.map (trans ctx) args)
+  | EStruct { name; fields } ->
+    let fields_name, fields = List.split (StructField.Map.bindings fields) in
+    Ast.monad_mbind_cont
+      (fun xs ->
+        let fields =
+          ListLabels.fold_right2 fields_name
+            (List.map (fun x -> Ast.monad_return ~mark (Expr.evar x mark)) xs)
+            ~f:StructField.Map.add ~init:StructField.Map.empty
+        in
+        Ast.monad_return ~mark (Expr.estruct name fields mark))
+      (List.map (trans ctx) fields)
+      ~mark
+  | EIfThenElse { cond; etrue; efalse } ->
+    (* There is two different encoding of the if then else. The first one is to
+       consider it as if it is an operator. Hence, if one of the branches is an
+       EmptyError, then it propagate to the final result of the expression. The
+       second one is [<<|cond |- a >, <|not cond |- b>| false :- empty>]. The
+       second semantics is redondant with exising default terms, but is the one
```

Comment:
What is the goal of this comment? Yes we could have decided not to have `EIfThenElse` at the Dcalc level and encode them using defaults, but as you mention it yields a different semantics because `EEmptyError` are propagated differently. Having `EIfThenElse` that behave as real `EIfThenElse` is a feature, not a bug :)